### PR TITLE
🎉 Amazon Sp Api: adds GET_V2_SETTLEMENT_REPORT_DATA_FLAT_FILE_V2 incremental report

### PR DESCRIPTION
**Pull request by VenezianoMauro**
## What
Adds GET_V2_SETTLEMENT_REPORT_DATA_FLAT_FILE_V2. 

## How
The new class FlatFileSettlementV2ReportsV2 uses the existing ReportsAmazonSPStream class plus the IncrementalMixin.
To make it incremental, the cursor is set to track the createdSince parameter of the request.  

## Recommended reading order
1. `x.java`
2. `y.python`

## 🚨 User Impact 🚨
Are there any breaking changes? What is the end result perceived by the user? If yes, please merge this PR with the 🚨🚨 emoji so changelog authors can further highlight this if needed.

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [x] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

### CODE REVIEW

**Comment by VenezianoMauro** [comment]
image

**Comment by marcosmarxm** [comment]
Thanks for the contribution @VenezianoMauro someone will review during the week.

**Comment by marcosmarxm** [comment]
@sh4sh do you mind reviewing this one?

@VenezianoMauro sorry the delay.

File: airbyte-integrations/connectors/source-amazon-seller-partner/integration_tests/configured_catalog_settlement_reports_v2.json [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+        "supported_sync_modes": ["full_refresh", "incremental"],
+        "source_defined_cursor": true,
+        "default_cursor_field": ["createdSince"]
+      },
+      "sync_mode": "incremental",
+      "destination_sync_mode": "append",
+      "cursor_field": ["createdSince"]
+    }
+  ]
+}
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-amazon-seller-partner/integration_tests/configured_catalog_settlement_reports_v2.json:16]: Can you add this to `configured_catalog` to run  tests too?

File: airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/streams.py [hunk 2]
```diff
@@ -1163,3 +1165,160 @@ def stream_slices(
...
+        
+        if end_date < strict_start_date:
+            end_date = strict_start_date
+
+        if end_date < start_date:
+            return []
+
+        params = {
+            "reportTypes": self.name,
+            "pageSize": 100,
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/streams.py:1237]:
```suggestion
            "pageSize": self.page_size,
```
and create the variable

File: airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/streams.py [hunk 3]
```diff
@@ -1163,3 +1165,160 @@ def stream_slices(
...
+                self._cursor_value =  new_cursor_value
+                logger.info(f"Updating cursor value for {self.name} stream to {self._cursor_value}")
+            
+
+        elif is_fatal:
+            raise Exception(f"The report for stream '{self.name}' was aborted due to a fatal error")
+        elif is_cancelled:
+            logger.warn(f"The report for stream '{self.name}' was cancelled or there is no data to return")
+        else:
+            raise Exception(f"Unknown response for stream `{self.name}`. Response body {report_payload}")
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/streams.py:1324]: EOF

**Review by marcosmarxm — CHANGES_REQUESTED** [review]
Left some comments.

File: airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/streams.py [hunk 4]
```diff
@@ -1163,3 +1165,160 @@ def stream_slices(
...
+        report_id = self._create_report(sync_mode, cursor_field, stream_slice, stream_state)["reportId"]
+
+        # create and retrieve the report
+        while not is_processed and seconds_waited < self.max_wait_seconds:
+            report_payload = self._retrieve_report(report_id=report_id)
+            seconds_waited = (pendulum.now("utc") - start_time).seconds
+            is_processed = report_payload.get("processingStatus") not in ["IN_QUEUE", "IN_PROGRESS"]
+            is_done = report_payload.get("processingStatus") == "DONE"
+            is_cancelled = report_payload.get("processingStatus") == "CANCELLED"
+            is_fatal = report_payload.get("processingStatus") == "FATAL"
```
[the remark this answers is not in the document]
Comment (VenezianoMauro) [reply airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/streams.py:1297]: i implemented the class in "constants". Is it okey?

File: airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/streams.py [hunk 5]
```diff
@@ -1163,3 +1166,163 @@ def stream_slices(
...
+        # create and retrieve the report
+        while not is_processed and seconds_waited < self.max_wait_seconds:
+            report_payload = self._retrieve_report(report_id=report_id)
+            seconds_waited = (pendulum.now("utc") - start_time).seconds
+            processing_status = report_payload.get("processingStatus")
+
+            is_processed = ProcessingStatus.is_processed(processing_status)
+            is_done = processing_status == ProcessingStatus.DONE
+            is_cancelled = processing_status == ProcessingStatus.CANCELLED
+            is_fatal = processing_status == ProcessingStatus.FATAL
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/streams.py:1299]: Please refactor this to status and not to have one variable to each status.

File: airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/streams.py [hunk 6]
```diff
@@ -1163,3 +1166,163 @@ def stream_slices(
...
+            
+            time.sleep(self.sleep_seconds)
+        
+        if is_done:
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/streams.py:1303]:
```suggestion
        if status == ProcessingStatus.DONE:
```

File: airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/streams.py [hunk 7]
```diff
@@ -1163,3 +1166,163 @@ def stream_slices(
...
+                yield r
+
+            new_cursor_value = stream_slice.get("createdUntil")
+            if (self._cursor_value is None) or (pendulum.parse(new_cursor_value) > pendulum.parse(self._cursor_value)):
+                self._cursor_value =  new_cursor_value
+                logger.info(f"Updating cursor value for {self.name} stream to {self._cursor_value}")
+            
+
+        elif is_fatal:
+            raise Exception(f"The report for stream '{self.name}' was aborted due to a fatal error")
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/streams.py:1324]: What I said here is, if there is a fatal error from the API please log it to the user understand what caused it

File: airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/streams.py [hunk 8]
```diff
@@ -1163,3 +1166,163 @@ def stream_slices(
...
+        elif is_cancelled:
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/streams.py:1325]:
```suggestion
        elif status == ProcessingStatus.CANCELLED:
```

File: airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/streams.py (hunk 5 above)
Comment (marcosmarxm) [reply airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/streams.py:1299]: You can use the `processing_status`

**Review by marcosmarxm — CHANGES_REQUESTED** [review]
Some comments.

**Comment by sajarin** [comment]
Hey @VenezianoMauro sorry for the delay in review here, it's been a while since we heard from you and there have been some comments left regarding this PR. 

If you have a chance, please create a new PR that rebases the master branch with the requested changes. We'd love to get this report into our connector if possible and we appreciate your contributions here!

**Pull request closed without merging**